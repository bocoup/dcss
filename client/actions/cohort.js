import {
  // CREATE_COHORT,
  CREATE_COHORT_SUCCESS,
  CREATE_COHORT_ERROR,
  SET_COHORT_SUCCESS,
  SET_COHORT_ERROR,
  // GET_COHORT,
  GET_COHORT_SUCCESS,
  GET_COHORT_ERROR,
  // GET_COHORT_PARTICIPANTS,
  GET_COHORT_PARTICIPANTS_SUCCESS,
  GET_COHORT_PARTICIPANTS_ERROR,
  // GET_COHORT_RUN_DATA_SUCCESS,
  GET_COHORT_RUN_DATA_ERROR,
  // GET_ALL_COHORTS,
  GET_ALL_COHORTS_SUCCESS,
  GET_ALL_COHORTS_ERROR,
  // GET_USER_COHORTS,
  GET_USER_COHORTS_SUCCESS,
  GET_USER_COHORTS_ERROR,
  // LINK_COHORT_TO_RUN,
  LINK_RUN_TO_COHORT_SUCCESS,
  LINK_RUN_TO_COHORT_ERROR,
  // SET_COHORT_USER_ROLE,
  SET_COHORT_USER_ROLE_SUCCESS,
  SET_COHORT_USER_ROLE_ERROR
} from './types';

export const createCohort = ({ name }) => async dispatch => {
  try {
    const res = await (await fetch('/api/cohort', {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ name })
    })).json();

    if (res.error) {
      throw res;
    }
    const { cohort } = res;

    dispatch({ type: CREATE_COHORT_SUCCESS, cohort });
    // return the cohort to the promise action for redirection purposes
    return cohort;
  } catch (error) {
    dispatch({ type: CREATE_COHORT_ERROR, error });
    return null;
  }
};

export const setCohort = cohort => async dispatch => {
  // dispatch({ type: SET_COHORT, cohort });
  try {
    const { scenarios } = cohort;
    const body = JSON.stringify({
      scenarios
    });
    await (await fetch(`/api/cohort/${cohort.id}/scenarios`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body
    })).json();
    dispatch({ type: SET_COHORT_SUCCESS, cohort });
  } catch (error) {
    dispatch({ type: SET_COHORT_ERROR, error });
    return null;
  }
};

export const getCohort = id => async dispatch => {
  if (Number.isNaN(id)) {
    return;
  }
  try {
    const res = await (await fetch(`/api/cohort/${id}`)).json();
    if (res.error) {
      throw res;
    }
    const { cohort } = res;
    dispatch({ type: GET_COHORT_SUCCESS, cohort });
    // return the cohort to the promise action for redirection purposes
    return cohort;
  } catch (error) {
    dispatch({ type: GET_COHORT_ERROR, error });
    return null;
  }
};

export const getCohorts = () => async dispatch => {
  try {
    const response = await (await fetch('/api/cohort/my')).json();

    if (response.error) {
      throw response;
    }
    const { cohorts } = response;
    dispatch({ type: GET_USER_COHORTS_SUCCESS, cohorts });
  } catch (error) {
    dispatch({ type: GET_USER_COHORTS_ERROR, error });
    return null;
  }
};

export const getAllCohorts = () => async dispatch => {
  try {
    const res = await (await fetch('/api/cohort/all')).json();
    if (res.error) {
      throw res;
    }
    const { cohorts } = res;
    dispatch({ type: GET_ALL_COHORTS_SUCCESS, cohorts });
    return cohorts;
  } catch (error) {
    dispatch({ type: GET_ALL_COHORTS_ERROR, error });
    return null;
  }
};

export const getCohortParticipants = id => async dispatch => {
  try {
    const res = await (await fetch(`/api/cohort/${id}`)).json();

    if (res.error) {
      throw res;
    }
    const {
      cohort,
      cohort: { users }
    } = res;
    // Dispatch the entire "cohort", but only the "users" property
    // will be used in the reducer.
    dispatch({ type: GET_COHORT_PARTICIPANTS_SUCCESS, cohort });
    // return the cohort to the promise action for redirection purposes
    return users;
  } catch (error) {
    dispatch({ type: GET_COHORT_PARTICIPANTS_ERROR, error });
    return null;
  }
};

export const getCohortData = (
  cohort_id,
  participant_id,
  scenario_id
) => async dispatch => {
  try {
    const endpoint = scenario_id
      ? `/api/cohort/${cohort_id}/scenario/${scenario_id}`
      : `/api/cohort/${cohort_id}/participant/${participant_id}`;

    const res = await (await fetch(endpoint)).json();

    if (res.error) {
      throw res;
    }

    const { prompts, responses } = res;

    // This dispatch does not update internal state, since
    // such state would be thrown away:
    // dispatch({ type: GET_COHORT_RUN_DATA_SUCCESS, prompts, responses });
    //
    return { prompts, responses };
  } catch (error) {
    dispatch({ type: GET_COHORT_RUN_DATA_ERROR, error });
    return null;
  }
};

export const linkRunToCohort = (cohort_id, run_id) => async dispatch => {
  try {
    const res = await (await fetch(
      `/api/cohort/${cohort_id}/run/${run_id}`
    )).json();
    if (res.error) {
      throw res;
    }
    const { cohort } = res;
    dispatch({ type: LINK_RUN_TO_COHORT_SUCCESS });
    dispatch({ type: GET_COHORT_SUCCESS, cohort });
    return cohort;
  } catch (error) {
    dispatch({ type: LINK_RUN_TO_COHORT_ERROR, error });
    return null;
  }
};
// This is used to
export const linkUserToCohort = (cohort_id, role) => async dispatch => {
  try {
    const users = await (await fetch(
      `/api/cohort/${cohort_id}/join/${role}`
    )).json();
    if (users.error) {
      throw users;
    }
    dispatch({ type: SET_COHORT_USER_ROLE_SUCCESS, users });
  } catch (error) {
    dispatch({ type: SET_COHORT_USER_ROLE_ERROR, error });
    return null;
  }
};

export const addCohortUserRole = (
  cohort_id,
  user_id,
  role
) => async dispatch => {
  try {
    const body = JSON.stringify({
      cohort_id,
      user_id,
      roles: [role]
    });
    const result = await (await fetch('/api/cohort/${cohort_id}/roles/add', {
      headers: {
        'Content-Type': 'application/json'
      },
      method: 'POST',
      body
    })).json();

    const { cohort } = result;
    dispatch({ type: SET_COHORT_USER_ROLE_SUCCESS });
    dispatch({ type: GET_COHORT_SUCCESS, cohort });
    return result;
  } catch (error) {
    dispatch({ type: SET_COHORT_USER_ROLE_ERROR, error });
    return null;
  }
};

export const deleteCohortUserRole = (
  cohort_id,
  user_id,
  role
) => async dispatch => {
  try {
    const body = JSON.stringify({
      cohort_id,
      user_id,
      roles: [role]
    });
    const result = await (await fetch('/api/cohort/${cohort_id}/roles/delete', {
      headers: {
        'Content-Type': 'application/json'
      },
      method: 'POST',
      body
    })).json();
    const { cohort } = result;
    dispatch({ type: SET_COHORT_USER_ROLE_SUCCESS });
    dispatch({ type: GET_COHORT_SUCCESS, cohort });
    return result;
  } catch (error) {
    dispatch({ type: SET_COHORT_USER_ROLE_ERROR, error });
    return null;
  }
};
