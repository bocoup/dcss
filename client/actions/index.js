import {
    LOG_IN,
    LOG_OUT,
    SET_RUN,
    SET_SCENARIO,
    SET_SCENARIOS,
    SET_SLIDES,
    SET_USERS
} from './types';

export const logIn = username => ({
    type: LOG_IN,
    payload: {
        username,
        isLoggedIn: true
    }
});

export const logOut = username => ({
    type: LOG_OUT,
    payload: {
        username,
        isLoggedIn: false
    }
});

export const setRun = payload => ({
    type: SET_RUN,
    payload
});

export const setScenario = payload => ({
    type: SET_SCENARIO,
    payload
});

export const setScenarios = payload => ({
    type: SET_SCENARIOS,
    payload
});

export const setSlides = payload => ({
    type: SET_SLIDES,
    payload
});

export const setUsers = payload => ({
    type: SET_USERS,
    payload
});
