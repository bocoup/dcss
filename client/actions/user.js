import Crypto from 'crypto-js';
import {
  GET_SESSION_SUCCESS,
  GET_SESSION_ERROR,
  GET_USER_SUCCESS,
  GET_USER_ERROR,
  SET_USER_SUCCESS,
  SET_USER_ERROR
} from './types';

export let getUser = () => async dispatch => {
  try {
    const res = await (await fetch('/api/auth/me')).json();

    if (res.error) {
      throw res;
    }
    const { user } = res;

    dispatch({ type: GET_USER_SUCCESS, user });
    return user;
  } catch (error) {
    dispatch({ type: GET_USER_ERROR, error });
    return null;
  }
};

export let getSession = () => async dispatch => {
  try {
    const res = await (await fetch('/api/auth/session')).json();

    if (res.error) {
      throw res;
    }
    const { user } = res;

    dispatch({ type: GET_SESSION_SUCCESS, user });
    return user;
  } catch (error) {
    dispatch({ type: GET_SESSION_ERROR, error });
    return null;
  }
};

export let setUser = data => async dispatch => {
  try {
    if (Object.values(data).length) {
      const body = JSON.stringify(data);
      const res = await (
        await fetch('/api/auth/update', {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json'
          },
          body
        })
      ).json();

      if (res.error) {
        throw res;
      }
      const { user } = res;

      dispatch({ type: SET_USER_SUCCESS, user });
      return user;
    }
  } catch (error) {
    dispatch({ type: SET_USER_ERROR, error });
    return null;
  }
};

export let resetPassword = data => async dispatch => {
  try {
    if (Object.values(data).length) {
      data.email = Crypto.AES.encrypt(data.email, SESSION_SECRET).toString();
      const body = JSON.stringify(data);
      const res = await (
        await fetch('/api/auth/reset', {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json'
          },
          body
        })
      ).json();

      if (res.error) {
        throw res;
      }
      const { reset } = res;
      return reset;
    }
  } catch (error) {
    return null;
  }
};
