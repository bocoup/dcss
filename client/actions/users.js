import { GET_USERS_SUCCESS, GET_USERS_ERROR } from './types';

export const setUsers = users => ({
  type: GET_USERS_SUCCESS,
  users
});

export const getUsers = () => async dispatch => {
  try {
    const { users } = await (await fetch('/api/roles/all')).json();
    dispatch({ type: GET_USERS_SUCCESS, users });
    return users;
  } catch (error) {
    dispatch({ type: GET_USERS_ERROR, error });
    return error;
  }
};
