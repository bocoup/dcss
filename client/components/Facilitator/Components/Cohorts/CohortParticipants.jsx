import React from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router';
import PropTypes from 'prop-types';
import {
  // NOTE: The checkbox is temporarily disabled
  // Checkbox,
  Container,
  // Dimmer,
  Icon,
  Input,
  Menu,
  Ref,
  Popup,
  Table
} from 'semantic-ui-react';
import _ from 'lodash';
import {
  getCohort,
  getCohortParticipants,
  setCohort
} from '@client/actions/cohort';
import EditorMenu from '@components/EditorMenu';
import ClickableTableCell from '@components/ClickableTableCell';
import ConfirmAuth from '@components/ConfirmAuth';
import scrollIntoView from '@components/util/scrollIntoView';
import './Cohort.css';

export class CohortParticipants extends React.Component {
  constructor(props) {
    super(props);

    let {
      params: { id }
    } = this.props.match;

    if (!id && this.props.id) {
      id = this.props.id;
    }

    this.state = {
      search: '',
      cohort: {
        id,
        users: []
      }
    };

    this.refreshInterval = null;

    // This is used as a back up copy of
    // participants when the list is filtered
    // by searching.
    this.participants = [];
    this.tableBody = React.createRef();
    this.onParticipantCheckboxClick = this.onParticipantCheckboxClick.bind(
      this
    );
    this.onParticipantSearchChange = this.onParticipantSearchChange.bind(this);
  }

  async componentDidMount() {
    const {
      cohort: { id }
    } = this.state;

    const cohort = await this.props.getCohort(Number(id));

    this.setState({
      cohort
    });

    this.refreshInterval = setInterval(async () => {
      const { search, cohort } = this.state;
      if (!search) {
        cohort.users = await this.props.getCohortParticipants(Number(id));
        this.participants = cohort.users.slice();
        this.setState({ cohort });
      }
    }, 1000);
  }

  async componentWillUnmount() {
    clearInterval(this.refreshInterval);
  }

  onParticipantCheckboxClick() {}

  scrollIntoView() {
    scrollIntoView(this.tableBody.current.node.firstElementChild);
  }

  onParticipantSearchChange(event, { value }) {
    const { participants } = this;
    const { cohort } = this.props;

    const escapedRegExp = new RegExp(_.escapeRegExp(value), 'i');

    const users = participants.filter(scenario => {
      if (escapedRegExp.test(scenario.username)) {
        return true;
      }

      if (escapedRegExp.test(scenario.email)) {
        return true;
      }
      return false;
    });

    this.setState({
      search: value,
      cohort: {
        ...cohort,
        users
      }
    });
  }

  render() {
    const { onClick } = this.props;
    const { cohort } = this.state;
    const { onParticipantSearchChange } = this;
    // NOTE: The checkbox is temporarily disabled
    // const { onParticipantCheckboxClick } = this;

    return (
      <Container fluid className="cohort__table-container">
        <EditorMenu
          type="cohort participants"
          items={{
            left: [
              <Menu.Item
                key="menu-item-cohort-scenarios"
                className="editormenu__padding"
                name="Scenarios in this Cohort"
                onClick={this.scrollIntoView}
              >
                <Icon.Group className="editormenu__icon-group">
                  <Icon name="group" />
                </Icon.Group>
                Cohort Participants ({this.props.cohort.users.length})
              </Menu.Item>
            ],
            right: [
              <ConfirmAuth
                key="menu-item-cohort-scenarios-search"
                requiredPermission="edit_scenarios_in_cohort"
              >
                <Menu.Menu position="right">
                  <Menu.Item
                    key="menu-item-search-accounts"
                    name="Search cohort participants"
                    className="editormenu__padding"
                  >
                    <Input
                      icon="search"
                      placeholder="Search..."
                      onChange={onParticipantSearchChange}
                    />
                  </Menu.Item>
                </Menu.Menu>
              </ConfirmAuth>
            ]
          }}
        />
        <Table
          fixed
          striped
          selectable
          role="grid"
          aria-labelledby="header"
          className="cohort__table--constraints"
          unstackable
        >
          <Table.Header className="cohort__table-thead-tbody-tr">
            <Table.Row>
              <Table.HeaderCell className="cohort__table-cell-first">
                {' '}
              </Table.HeaderCell>
              <Table.HeaderCell>Username</Table.HeaderCell>
              <Table.HeaderCell>Role</Table.HeaderCell>
              <Table.HeaderCell className="cohort__table-cell-content">
                Email
              </Table.HeaderCell>
            </Table.Row>
          </Table.Header>
          <Ref innerRef={node => (this.tableBody = node)}>
            <Table.Body className="cohort__scrolling-tbody">
              {cohort.users.length ? (
                cohort.users.map((user, index) => {
                  const onClickAddTab = (event, data) => {
                    onClick(event, {
                      ...data,
                      type: 'participant',
                      source: user
                    });
                  };

                  return (
                    <Table.Row
                      key={`participants-row-${index}`}
                      className="cohort__table-thead-tbody-tr"
                      style={{ cursor: 'pointer' }}
                    >
                      {/*
                                            <Table.Cell
                                                key={`participants-cell-checkbox-${index}`}
                                                className="cohort__table-cell-first"
                                            >
                                                <Popup
                                                    content="Adding participants is not available in this version of Cohorts"
                                                    trigger={
                                                        <Checkbox
                                                            disabled
                                                            key={`participants-checkbox-${index}`}
                                                            value={user.id}
                                                            onClick={
                                                                onParticipantCheckboxClick
                                                            }
                                                        />
                                                    }
                                                />
                                            </Table.Cell>
                                            */}

                      <ConfirmAuth requiredPermission="view_all_data">
                        <Popup
                          content="View cohort reponses from this participant"
                          trigger={
                            <ClickableTableCell
                              className="cohort__table-cell-first"
                              display={<Icon name="file alternate outline" />}
                              onClick={onClickAddTab}
                            />
                          }
                        />
                      </ConfirmAuth>
                      <Table.Cell>{user.username}</Table.Cell>
                      <Table.Cell>{user.role}</Table.Cell>
                      <Table.Cell className="cohort__table-cell-content">
                        {user.email}
                      </Table.Cell>
                    </Table.Row>
                  );
                })
              ) : (
                <Table.Row
                  key={`row-empty-results`}
                  className="cohort__table-thead-tbody-tr"
                >
                  <Table.Cell>No participants match your search</Table.Cell>
                </Table.Row>
              )}
            </Table.Body>
          </Ref>
        </Table>
      </Container>
    );
  }
}

CohortParticipants.propTypes = {
  cohort: PropTypes.shape({
    id: PropTypes.any,
    name: PropTypes.string,
    role: PropTypes.string,
    runs: PropTypes.array,
    scenarios: PropTypes.array,
    users: PropTypes.array
  }),
  participants: PropTypes.array,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired,
  id: PropTypes.any,
  match: PropTypes.shape({
    path: PropTypes.string,
    params: PropTypes.shape({
      id: PropTypes.node
    }).isRequired
  }).isRequired,
  onClick: PropTypes.func,
  getCohort: PropTypes.func,
  getCohortParticipants: PropTypes.func,
  setCohort: PropTypes.func,
  scenarios: PropTypes.array
};

const mapStateToProps = state => {
  const { currentCohort: cohort } = state.cohort;
  const { users: participants } = cohort;
  const { scenarios } = state.scenario;
  return { cohort, participants, scenarios };
};

const mapDispatchToProps = dispatch => ({
  getCohort: id => dispatch(getCohort(id)),
  getCohortParticipants: id => dispatch(getCohortParticipants(id)),
  setCohort: params => dispatch(setCohort(params))
});

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(CohortParticipants)
);
