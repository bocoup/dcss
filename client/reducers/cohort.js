import {
  CREATE_COHORT_SUCCESS,
  GET_COHORT_PARTICIPANTS_SUCCESS,
  GET_COHORT_SUCCESS,
  GET_ALL_COHORTS_SUCCESS,
  GET_USER_COHORTS_SUCCESS,
  SET_COHORT_SUCCESS,
  SET_COHORT_USER_ROLE_SUCCESS
} from '../actions/types';

const cohortInitialState = {
  id: null,
  created_at: '',
  name: '',
  runs: [],
  scenarios: [],
  users: []
};

export const cohort = (state = cohortInitialState, action) => {
  const { type, cohort, users } = action;

  if (type === SET_COHORT_SUCCESS || type === GET_COHORT_SUCCESS) {
    return {
      ...state,
      ...cohort
    };
  }

  if (type === CREATE_COHORT_SUCCESS) {
    // console.log('CREATED: ', { ...cohort });
    return {
      ...state,
      ...cohort,
      role: 'owner'
    };
  }

  if (type === GET_COHORT_PARTICIPANTS_SUCCESS) {
    return {
      ...state,
      users: cohort.users
    };
  }

  if (type === SET_COHORT_USER_ROLE_SUCCESS) {
    if (users) {
      return {
        ...state,
        users
      };
    } else {
      return {
        ...state
      };
    }
  }

  return state;
};

export const cohorts = (state = [], action) => {
  const { type, cohorts } = action;
  if (type === GET_USER_COHORTS_SUCCESS || type === GET_ALL_COHORTS_SUCCESS) {
    return cohorts.slice();
  }

  return state;
};

export const cohortsById = (state = {}, action) => {
  const { type, cohorts } = action;

  if (type === GET_USER_COHORTS_SUCCESS || type === GET_ALL_COHORTS_SUCCESS) {
    return cohorts.reduce((accum, cohort) => {
      accum[cohort.id] = cohort;
      return accum;
    }, {});
  }

  return state;
};
