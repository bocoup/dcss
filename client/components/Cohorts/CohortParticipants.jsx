import React from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';
import {
  // NOTE: The checkbox is temporarily disabled
  // Checkbox,
  Container,
  // Dimmer,
  Icon,
  Input,
  Menu,
  Popup,
  Ref,
  Table
} from '@components/UI';
import escapeRegExp from 'lodash.escaperegexp';
import Identity from '@utils/Identity';
import Storage from '@utils/Storage';
import { getCohort, setCohort } from '@actions/cohort';
import EditorMenu from '@components/EditorMenu';
import Username from '@components/User/Username';
import UsersTable from '@components/Admin/UsersTable';
import ConfirmAuth from '@components/ConfirmAuth';
import Loading from '@components/Loading';
import scrollIntoView from '@utils/scrollIntoView';
import { COHORT_ROLE_GROUPS } from '../Admin/constants';
import './Cohort.css';

const { facilitator, researcher } = COHORT_ROLE_GROUPS;

import Layout from '@utils/Layout';
const ROWS_PER_PAGE = 10;

export class CohortParticipants extends React.Component {
  constructor(props) {
    super(props);

    let {
      params: { id }
    } = this.props.match;

    if (!id && this.props.id) {
      id = this.props.id;
    }

    this.sessionKey = `cohort-participants/${id}`;
    let persisted = Storage.get(this.sessionKey);

    if (!persisted) {
      persisted = { refresh: false };
      Storage.set(this.sessionKey, persisted);
    }

    const { refresh } = persisted;

    this.state = {
      isReady: false,
      activePage: 1,
      refresh,
      search: '',
      cohort: {
        id,
        users: []
      }
    };

    this.refreshInterval = null;

    // This is used as a back up copy of
    // participants when the list is filtered
    // by searching.
    this.participants = [];
    this.tableBody = React.createRef();
    this.sectionRef = React.createRef();
    this.onPageChange = this.onPageChange.bind(this);
    this.onParticipantSearchChange = this.onParticipantSearchChange.bind(this);
    this.onParticipantRefreshChange = this.onParticipantRefreshChange.bind(
      this
    );
  }

  async fetchCohort() {
    const {
      cohort: { id }
    } = this.state;

    const cohort = await this.props.getCohort(Number(id));

    this.participants = cohort.users.slice();
    this.setState({
      isReady: true,
      cohort
    });
  }

  participantRefresh() {
    this.refreshInterval = setInterval(async () => {
      if (!this.state.search) {
        await this.fetchCohort();
      }
    }, 1000);
  }

  async componentDidMount() {
    await this.fetchCohort();

    if (this.state.refresh) {
      this.participantRefresh();
    }
  }

  async componentWillUnmount() {
    clearInterval(this.refreshInterval);
  }

  scrollIntoView() {
    scrollIntoView(this.tableBody.current.node.firstElementChild);
  }

  onParticipantSearchChange(event, { value }) {
    const { participants } = this;
    const { cohort } = this.props;

    if (value === '') {
      this.setState({
        activePage: 1,
        search: value,
        cohort: {
          ...cohort,
          users: participants
        }
      });
      return;
    }

    if (value.length < 3) {
      return;
    }

    const escapedRegExp = new RegExp(escapeRegExp(value), 'i');
    let users = participants.filter(participant => {
      // console.log(participant);
      if (escapedRegExp.test(participant.username)) {
        return true;
      }

      if (escapedRegExp.test(participant.email)) {
        return true;
      }

      if (escapedRegExp.test(participant.roles.join(','))) {
        return true;
      }

      return false;
    });

    if (!value) {
      users = this.participants;
    }

    this.setState({
      activePage: 1,
      search: value,
      cohort: {
        ...cohort,
        users
      }
    });
  }

  onParticipantRefreshChange() {
    let refresh = !this.state.refresh;
    this.setState({ refresh }, () => {
      if (!refresh) {
        clearInterval(this.refreshInterval);
      } else {
        this.participantRefresh();
      }
      Storage.set(this.sessionKey, this.state);
    });
  }

  onPageChange(event, { activePage }) {
    this.setState({
      activePage
    });
  }

  render() {
    const { authority, onClick, user, usersById } = this.props;
    const { activePage, cohort, isReady, refresh } = this.state;
    const {
      onPageChange,
      onParticipantRefreshChange,
      onParticipantSearchChange,
      scrollIntoView
    } = this;
    // Ensure that Facilitator access is applied even if the user just
    // became a facilitator and their session roles haven't updated.
    const isFacilitator =
      user.roles.includes('facilitator') || authority.isFacilitator;

    if (!isReady) {
      return <Loading />;
    }

    const refreshIcon = refresh ? 'play' : 'square';
    const refreshColor = refresh ? 'green' : 'red';
    const refreshLabel = refresh
      ? 'This list will refresh automatically'
      : 'This list will refresh when page is reloaded';

    const pages = Math.ceil(cohort.users.length / ROWS_PER_PAGE);
    const index = (activePage - 1) * ROWS_PER_PAGE;
    const users = cohort.users.slice(index, index + ROWS_PER_PAGE);
    const columns = {
      data: {
        className: 'c__table-cell-first c__hidden-on-mobile',
        content: ''
      },
      username: {
        className: 'users__col-large',
        content: 'User'
      },
      email: {
        className: 'users__col-large c__hidden-on-mobile',
        content: 'Email'
      }
    };

    const grantableRoles = {};

    if (user.roles.includes('facilitator') || user.is_super) {
      Object.assign(grantableRoles, facilitator);
    }

    if (user.roles.includes('researcher')) {
      Object.assign(grantableRoles, researcher);
    }

    Object.assign(columns, grantableRoles);

    const rows = users.reduce((accum, cohortUser) => {
      // username will never be empty, but email might be.
      const onClickAddTab = (event, data) => {
        onClick(event, {
          ...data,
          type: 'participant',
          source: cohortUser
        });
      };
      const key = Identity.key(cohortUser);

      const trigger = (
        <Table.Cell.Clickable
          className="c__table-cell-first c__hidden-on-mobile"
          key={`p-${key}`}
          content={<Icon name="file alternate outline" />}
          onClick={onClickAddTab}
        />
      );
      const popup = (
        <ConfirmAuth
          key={`c-${key}`}
          isAuthorized={isFacilitator}
          requiredPermission="view_all_data"
        >
          <Popup
            inverted
            size="tiny"
            content="View cohort reponses from this participant"
            trigger={trigger}
          />
        </ConfirmAuth>
      );

      const { is_super } = (usersById && usersById[cohortUser.id]) || {};
      const username = <Username {...cohortUser} is_super={is_super} />;

      const usernameCell = cohortUser.roles.includes('owner') ? (
        <Table.Cell key={`u-${key}`}>{username} (owner)</Table.Cell>
      ) : (
        <Table.Cell key={`u-${key}`}>{username}</Table.Cell>
      );

      accum[cohortUser.id] = [
        popup,
        usernameCell,
        <Table.Cell className="c__hidden-on-mobile" key={`e-${key}`}>
          {cohortUser.email}
        </Table.Cell>
      ];

      return accum;
    }, {});

    const usersTableProps = {
      activePage,
      cohort,
      columns,
      grantableRoles,
      onPageChange,
      pages,
      rows
    };

    const left = [
      <Menu.Item.Tabbable
        key="menu-item-cohort-participants"
        name="Participants in this Cohort"
        onClick={scrollIntoView}
      >
        <Icon.Group className="em__icon-group-margin">
          <Icon name="group" />
        </Icon.Group>
        Participants ({this.props.cohort.users.length})
      </Menu.Item.Tabbable>,
      Layout.isNotForMobile() ? (
        <Menu.Item.Tabbable
          key="menu-item-cohort-participants"
          name="Control participant list refresh"
          onClick={onParticipantRefreshChange}
        >
          <Icon.Group className="em__icon-group-margin">
            <Icon name="refresh" />
            <Icon corner="top right" name={refreshIcon} color={refreshColor} />
          </Icon.Group>
          {refreshLabel}
        </Menu.Item.Tabbable>
      ) : null
    ];

    const right = [
      <Menu.Menu key="menu-menu-search-cohort-participants" position="right">
        <Menu.Item.Tabbable
          key="menu-item-search-cohort-participants"
          name="Search cohort participants"
        >
          <Input
            icon="search"
            placeholder="Search..."
            onChange={onParticipantSearchChange}
          />
        </Menu.Item.Tabbable>
      </Menu.Menu>
    ];
    const editorMenu = (
      <Ref innerRef={node => (this.sectionRef = node)}>
        <EditorMenu type="cohort participants" items={{ left, right }} />
      </Ref>
    );
    return (
      <Container fluid className="c__table-container">
        {editorMenu}
        <UsersTable {...usersTableProps} />
      </Container>
    );
  }
}

CohortParticipants.propTypes = {
  authority: PropTypes.object,
  cohort: PropTypes.shape({
    id: PropTypes.any,
    name: PropTypes.string,
    role: PropTypes.string,
    runs: PropTypes.array,
    scenarios: PropTypes.array,
    users: PropTypes.array
  }),
  participants: PropTypes.array,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired,
  id: PropTypes.any,
  match: PropTypes.shape({
    path: PropTypes.string,
    params: PropTypes.shape({
      id: PropTypes.node
    }).isRequired
  }).isRequired,
  onClick: PropTypes.func,
  getCohort: PropTypes.func,
  setCohort: PropTypes.func,
  scenarios: PropTypes.array,
  user: PropTypes.object,
  usersById: PropTypes.object
};

const mapStateToProps = state => {
  const { cohort, user, usersById } = state;
  const { users: participants } = cohort;
  const { scenarios } = state.scenario;
  return { cohort, participants, scenarios, user, usersById };
};

const mapDispatchToProps = dispatch => ({
  getCohort: id => dispatch(getCohort(id)),
  setCohort: params => dispatch(setCohort(params))
});

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(CohortParticipants)
);
