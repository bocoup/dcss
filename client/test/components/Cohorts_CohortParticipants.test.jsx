import {
  fetchImplementation,
  mounter,
  reduxer,
  snapshot,
  state
} from '../bootstrap';
import { unmountComponentAtNode } from 'react-dom';
import { cleanup, fireEvent, render } from '@testing-library/react';
import CohortParticipants from '../../components/Cohorts/CohortParticipants.jsx';

import {
  GET_COHORT_SUCCESS,
  GET_USER_SUCCESS,
  GET_USERS_SUCCESS
} from '../../actions/types';
import * as cohortActions from '../../actions/cohort';
import * as userActions from '../../actions/user';
import * as usersActions from '../../actions/users';
jest.mock('../../actions/cohort');
jest.mock('../../actions/user');
jest.mock('../../actions/users');

beforeAll(() => {
  (window || global).fetch = jest.fn();
});

let container = null;
beforeEach(() => {
  container = document.createElement('div');
  container.setAttribute('id', 'root');
  document.body.appendChild(container);

  fetchImplementation(fetch);

  cohortActions.getCohort = jest.fn();
  cohortActions.getCohort.mockImplementation(() => async dispatch => {
    const cohort = {
      id: 2,
      created_at: '2020-08-31T14:01:08.656Z',
      name: 'A New Cohort That Exists Within Inline Props',
      runs: [],
      scenarios: [],
      users: [
        {
          id: 2,
          email: 'owner@email.com',
          username: 'owner',
          cohort_id: 2,
          roles: ['owner', 'facilitator'],
          is_anonymous: false,
          is_owner: true
        }
      ],
      roles: ['owner', 'facilitator'],
      usersById: {
        2: {
          id: 2,
          email: 'owner@email.com',
          username: 'owner',
          cohort_id: 2,
          roles: ['owner', 'facilitator'],
          is_anonymous: false,
          is_owner: true
        }
      }
    };
    dispatch({
      type: GET_COHORT_SUCCESS,
      cohort
    });
    return cohort;
  });
  usersActions.getUser = jest.fn();
  usersActions.getUser.mockImplementation(() => async dispatch => {
    const user = {
      id: 2,
      email: 'owner@email.com',
      username: 'owner',
      personalname: 'Owner Account',
      roles: ['owner'],
      is_owner: true,
      is_author: true,
      is_reviewer: false
    };
    dispatch({
      type: GET_USER_SUCCESS,
      user
    });
    return user;
  });
  usersActions.getUsers = jest.fn();
  usersActions.getUsers.mockImplementation(() => async dispatch => {
    const users = [
      {
        id: 2,
        email: 'owner@email.com',
        username: 'owner',
        personalname: 'Owner Account',
        roles: ['owner'],
        is_owner: true,
        is_author: true,
        is_reviewer: false
      }
    ];
    dispatch({
      type: GET_USERS_SUCCESS,
      users
    });
    return users;
  });
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

const sharedProps = {
  history: {
    push() {}
  }
};

test('CohortParticipants', () => {
  expect(CohortParticipants).toBeDefined();
});

test('Snapshot 1', () => {
  const props = {
    ...sharedProps,
    authority: {}
  };
  const mounted = mounter(reduxer(CohortParticipants, props, state))();
  expect(snapshot(mounted)).toMatchSnapshot();
});

/*{INJECTION}*/
