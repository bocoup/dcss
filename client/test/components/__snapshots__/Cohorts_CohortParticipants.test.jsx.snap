// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render 1 1 1`] = `
"<body>
  <div
    id=\\"root\\"
  />
  <div>
    <div
      class=\\"ui card loading__single-card\\"
      style=\\"width: 100%; height: 90%;\\"
    >
      <div
        class=\\"content loading__content\\"
      >
        <div
          class=\\"ui placeholder\\"
        >
          <div
            class=\\"square image\\"
          />
        </div>
      </div>
    </div>
  </div>
</body>"
`;
